### DISCUSSION TITLE
Way to protect panel ?

### DISCUSSION THREAD

**Discussion by FabQi** [RomanKubiak/ctrlr#617] [Q&A]
Hello

I plan to sell my panels...

Is there a "simple" way to "protect" then ? 

(Perhaps something with the "RSA Key used to signed the exported panel" ?)

Regards :-)
Fabrice

**Comment by damiensellier** [comment]
Hi Fabrice, RSA key option is not yet implemented, sorry for that. The button and menu are just placeholders for a function that is not yet under the hood unfortunately.
Best way to protect your plugin is via[ Pace iLok ](https://paceap.com/products/ilok-licensing-platform/)which will encrypt and protect your plugin via their Level 1 Protection (wrapper) or Level 2 [Fusion](https://paceap.com/products/fusion/) process. You'll need to get in touch with them and sign for a licence deal with them. You'll be able to use their sdk and produce highly secured softwares either via their iLok dongle key or with machine based software authorization or just encrypt the binaries.
 
Or if you want you can add [encryption modules from JUCE](https://docs.juce.com/master/group__juce__cryptography-encryption.html) and [JUCE serial challenge/response authorization](https://docs.juce.com/master/tutorial_online_unlock_status.html) to Ctrlr, classes are already available.

Also for info, the licences CTRLR is under (BSD 3& GNU) allow profitable sales but if I understand it requires [crediting](https://github.com/RomanKubiak/ctrlr/blob/master/LICENSE-BSD) CTRLR creator with a mention like "Made with Ctrlr by Roman Kubiak under blah bah licence" in your documentation and/or software about dialog box.
Same for JUCE (which ctrlr is based from) requiring their splash logo on startup unless you subscribe to a indie/pro [subscription plan](https://juce.com/get-juce/).
You also need to sign the licence agreement with steinberg to use their VST3 sdk as a binary. 
Nothing really complicated, but it's something to take in consideration. Just take a look at all that when you'll be ready to release your plugins.
Good luck! 
Damien

Edit : All above is for binaries including Ctrlr. If you plan to sell only the pannel, that's a different story. [You can do whatever you want. ](https://ctrlr.org/forums/topic/ctrlr-licence/#post-37755)But panels won't be encrypted and will just be compressed internally as gzip.

**Reply by monophreak** [reply]
Good Evening Damien, 

Thank you for all of your work on CTRLR. It is really appreciated and I hope all is well! 

I see you added Blowfish encryption for exported instances here: https://github.com/RomanKubiak/ctrlr/discussions/700

I have managed to successfully compile CtrlrX in Sequoia. Sadly, I cannot get this feature working on v5.6.33. 

// Define the BlowFish encryption key as string
String keyString = "yourkey"; // Replace with your actual key (security!) Added v5.6.31

The steps I have taken include installing Xcode and setting up JUCE to compile. I edit the key in the two locations from the source files. I'm not familiar with blowfish and I'm just putting in a random jumble of letters, numbers and symbols in place of "yourkey" which is maybe why it is failing. I place a restricted instance set of resources into a standard shell and no luck with the encryption. I tried a random key from here too: https://www.motorsportdiesel.com/tools/blowfish-salt/pma/

If you have any advice about what I may be doing wrong, please let me know.

Thanks,

Maxie

**Reply by damiensellier** [reply]
Hi, thanks!
You need to replace the 2 occurrences with the same custom key. It's in 2 functions on macOS only.

I recently updated the string to modulator something to blur the line if people read the binaries in an editor :)

You're almost there, you need the key both on encryption AND decryption. Same passkey obviously.

**Reply by damiensellier** [reply]
Note : since the panel resource is now encrypted in export (export instance), you cannot copy paste the exported resources in a stock CtrlrX bundle. It won't be able to read it and it's meant exactly for that. You need your panel encrypted with the exact same binary you built with Xcode with the custom key or the exported instance won't load and show a blank CtrlrX.

It's really basic, but it's just an extra protection to prevent smart hackers to just copy paste the custom panel resources from the bundle and paste it in a stock CtrlrX app or vst or else.

**Reply by damiensellier** [reply]
Can you please check you have PanelZBF in your bundle resource folder along with ResourcesZ ?

![Capture d’écran, le 2025-07-01 à 23 31 45](https://github.com/user-attachments/assets/03e85329-c5d8-4145-8ffd-bbd3310195d4)

**Reply by monophreak** [reply]
> Hi, thanks! You need to replace the 2 occurrences with the same custom key. It's in 2 functions on macOS only.
> 
> I recently updated the string to modulator something to blur the line if people read the binaries in an editor :)
> 
> You're almost there, you need the key both on encryption AND decryption. Same passkey obviously.

Thanks Damien, I saw the new modulator string in the code and thought that this was a nice touch.

I can confirm that I am placing the same key in both locations in the source code.

I export with a restricted instance from within CtrlrX and use JUCE 6.1.6 during the build. I was wondering if this feature is related to JUCE 7?

**Reply by monophreak** [reply]
> Note : since the panel resource is now encrypted in export (export instance), you cannot copy paste the exported resources in a stock CtrlrX bundle. It won't be able to read it and it's meant exactly for that. You need your panel encrypted with the exact same binary you built with Xcode with the custom key or the exported instance won't load and show a blank CtrlrX.
> 
> It's really basic, but it's just an extra protection to prevent smart hackers to just copy paste the custom panel resources from the bundle and paste it in a stock CtrlrX app or vst or else.

Basic Blowfish is fine for my purposes, when people request a copy of my latest panel (or sign up to my mailing list), I am happy to post out links to my latest version to try and keep some older, broken tech out of landfill. If people like my work, they are welcome to gift me a coffee :-)

**Reply by monophreak** [reply]
> Can you please check you have PanelZBF in your bundle resource folder along with ResourcesZ ?
> 
> ![Capture d’écran, le 2025-07-01 à 23 31 45](https://private-user-images.githubusercontent.com/55164385/461335390-03e85329-c5d8-4145-8ffd-bbd3310195d4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTE0NDU5NjgsIm5iZiI6MTc1MTQ0NTY2OCwicGF0aCI6Ii81NTE2NDM4NS80NjEzMzUzOTAtMDNlODUzMjktYzVkOC00MTQ1LThmZmQtYmJkMzMxMDE5NWQ0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzAyVDA4NDEwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWYwN2U0OGY3YWI1OWUzOTgxYzg2Y2QyMWJmOTVhMjJhMTIyNDg4YzBhZTcyMmJkYjM3NDg3YTFkNDIwNzRkYmImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.iz1jXyWtAsk5IIVus-RjRkAephuU-9Ss29e9EVre_WA)

I do! I can confirm it is there. It is a strange one this and I’m wondering if I am selecting the wrong export option (restricted instance) or whether it’s due to an older version of JUCE. I will play some more later on today.

Thank you so much for your support and fast replies Damien.

**Reply by monophreak** [reply]
> > Can you please check you have PanelZBF in your bundle resource folder along with ResourcesZ ?
> > ![Capture d’écran, le 2025-07-01 à 23 31 45](https://private-user-images.githubusercontent.com/55164385/461335390-03e85329-c5d8-4145-8ffd-bbd3310195d4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTE0NDU5NjgsIm5iZiI6MTc1MTQ0NTY2OCwicGF0aCI6Ii81NTE2NDM4NS80NjEzMzUzOTAtMDNlODUzMjktYzVkOC00MTQ1LThmZmQtYmJkMzMxMDE5NWQ0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA3MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNzAyVDA4NDEwOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWYwN2U0OGY3YWI1OWUzOTgxYzg2Y2QyMWJmOTVhMjJhMTIyNDg4YzBhZTcyMmJkYjM3NDg3YTFkNDIwNzRkYmImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.iz1jXyWtAsk5IIVus-RjRkAephuU-9Ss29e9EVre_WA)
> 
> I do! I can confirm it is there. It is a strange one this and I’m wondering if I am selecting the wrong export option (restricted instance) or whether it’s due to an older version of JUCE. I will play some more later on today.
> 
> Thank you so much for your support and fast replies Damien.

Ohhh, after re-reading, I am wondering if it is pilot error.

I have been creating my new export protected instance (app 1) and a standard export (app 2) to get the container for testing. I then swap resources between the two. Each exported app contains the CtrlrX app which I compiled and therefore has the key! So in effect, bad test!

Therefore, if someone used a different build of CtrlrX to create a standard export and then copied and pasted the resources from my restricted export into the resource section, this would fail because the CtrlrX inside the bundle is different and lacks the key. Would that be correct? I think that is logical and can’t wait to test this out!

**Reply by damiensellier** [reply]
Ok just to clarify a few things:

The encryption is just there to protect the panel datas embedded in the mac bundle from being copied to a stock CtrlrX binary.
It's really just like a zip with a password. the panel xml datas are always stored in the bundle resources. I added a blowfish encryption to protect panels with antipiracy mechanism such as ilok pace or using a top layer with a registration (user, password) system from people being able to just copy the panel and modify it to circumvent the security layer.

the password has nothing to do with a registration key showing up on startup for exemple when you open the exported instance.

 An "exported instance" is just a carbon copy of the original "empty" Ctrlr app, but with added resources (panel + panel resources such as jpg etc) inside the app bundle. When you export an instance of a panel from Ctrlr, the app "CtrlrX" copies itself to a new app "Panel_Name" and places the panel (zipped and encrypted) inside the "Panel_Name" app bundle /resources/ folder. Then it codesigns for local use because in a certain way you just "hacked" an app : its name is changed, the size is different and the info.plist is edited. If not re-codesigned, macOS will recognize it has a tampered app then will discard it to the blacklist.

before adding the blowfish encryption, the resources were copied without encryption. It means that if you did all the pace ilok process, or used a registration mechanism, you just had to copy the "exported instance" resources, un-gzip them, and you could easily remove in a text editor the protection script. With the extra layer of encryption, if you don't have the right password embedded inside your custom CtrlrX version executable (the 2 strings in CtrlrMac.app) it's impossible to read the panel resources to edit them.

Now, anyway all exported instances resources are encrypted, at least with the stock password, declared by "keyString".

_if someone used a different build of CtrlrX to create a standard export and then copied and pasted the resources from my restricted export into the resource section, this would fail because the CtrlrX inside the bundle is different and lacks the key. Would that be correct?_

Absolutely. Since "keyString" would be different, there's no way the panel could be decrypted and opened by the "stock" Ctrlr app.

An encrypted panel always goes together with a dedicated custom CtrlrX executable.

It's not a top level security but more like a deep layer of extra security for the developers who want to secure their commercial plugins with a registration process, ilok pace etc
If you just distribute open source freeware panels, just forget about it, it's useless.

BTW I'm going to implement the online registration module so it's going to be easier for developers to protect their commercial panels/plugins. 

https://github.com/damiensellier/CtrlrX/issues/115

Also as a reminder, you need to compile CtrlrX with the Projucer v6.0.8 or 6.1.3, and use exclusively the JUCE 6.0.8 libraries included inside /CtrlrX/JUCE/modules/ 

Good luck with your project Maxie

**Reply by damiensellier** [reply]
EDIT: reply above edited for clarity

**Reply by monophreak** [reply]
> 6.1.3

Thanks Damien, the online registration module sounds fantastic. 

I think things are working fine and I will test further. I created a new build of CtrlX from Projucer 6.1.3 and set the JUCE global paths to the JUCE folder inside the CtrlrX v5.6.33 master download and the global module paths to the modules inside the CtrlrX v5.6.33 download. I then ran a re-scan of all modules. I was not sure if there was another (or better) way to accomplish pointing to the libraries.

CtrlrX built correctly without errors. I downloaded the GitHub already compiled release of CtrlrX and created an export container with the default key string for that release. I swapped out the resources from my build with my key. Success! It did not load! At first, I thought it loaded but realised that when it could not load, CtrlrX would open and display the last project I was working on. Sometimes, this may have been the same as the test version but closing that down and repeating showed that nothing would actually load up.

I will play some more but this looks great. From my perspective, this enhancement is fantastic because I share the panels freely and do not worry about serials or ilok. Due to the way in which I share my work, I see no reason why anyone would want to hack it either. However, I still wanted to encrypt the panel contents to prevent unauthorised copying or editing of the code in my work because I depend on people coming back to my site and gifting a coffee if they wish to, to drive my next project forward. My next panel release represents about a year and half of my time in the making.

If there is a similar Windows solution to encrypt the resources folder on the horizon, that would be amazing. I plan to hold fire on Windows releases for a while or until a solution presents itself.

Thank you for all of your work on CtrlrX Damien, your enhancements have made the IDE very enjoyable to use.

**Reply by damiensellier** [reply]
That sounds great! This is indeed a very nice way to seal your plugin to prevent any mods or sharing aside your own channels.
Windows is different, it's not like with macOS and its bundle independent resources. On WIN, the panel and the resources during the export are embedded into the executable. So it's harder to mess with it, since it's in gzip format anyway, if you open the exe in an editor you won't see any readable code of your custom panel. 
I added the AAX support for protools to be released in the next version.
I'm also thinking about porting CtrlrX to AUv3 for iPad and adding kiosk mode for android tablets... 

Good luck and don't hesitate to reach me or the other members if you need anything.

Damien

**Reply by monophreak** [reply]
Hi Damien, this weekend, I tested the latest Blowfish encrpytion out on my own CtrlrX Intel builds and Silicon builds. Your work here is very impressive. Everything checks out! Thank you so much for this addition. You really are doing an amazing job in CtrlrX.

For the Silicon builds, I had a clean build with the share files and standalone builds. In case it helps anyone else here, I added the key settings that I used, to the forum post for Silicon builds found here ([link](https://github.com/RomanKubiak/ctrlr/issues/560)).

**Reply by damiensellier** [reply]
Thanks! Happy everything is working on your side. That's good news.
I'm taking a look at the online plugin registration/unlocking module with JUCE right now.

I think it's going to be available soon for CtrlrX...

**Comment by FabQi** [comment]
Hello

I'll put this in the todo list
Thank you again

Fabrice

**Reply by bijlevel** [reply]
If (a big if!) you make a profit, then at least consider a donation for Roman Kubiak...

**Comment by dobo365** [comment]
The other question is also: why do you want to spend time on that?
I mean...
How much will you sell your panel (I don't know what it is about): 10-20 EUR? Let's say max 50 EUR (but I think that at that price this is prohibitive)
How many do you think you will sell? This question is to put in parallel to: How many synth / units supported by your panel have been sold? How many of the users are still using it? How many of those would be interested in an editor?
Are you sure that the protection you will implement will be useful and not be boring for the users?
Taking into account that a lot (=most) users of our panels are willing to pay for what we did and the number of hours we spent, if you price it at the right level (= a price related to the cost of the HW but also affordable and reasonable), you will quickly discover that it is not worth spending time on locking the panels.
At least this is my humble opinion.
For sure, please be prepared that you cannot live with this "business" and that you will maybe just get enough money to purchase the actual unit...

**Reply by damiensellier** [reply]
I understand @dobo365  point of view. In my case I would be willing to pay some good money to have a robust and well made plugin to control some of my hardware. Pro studios would do it too. There's nothing better than being able to recall external gears from the DAW. Bricasti has a plugin for their reverb ([Liquidsonics M7 link ](https://www.liquidsonics.com/software/m7-link/)is 199usd), midi control is a huge sale argument for Tegeler Audio and [Wes Audio](https://www.youtube.com/watch?v=2F_QsOvlBqQ) too... It's really a thing now. 
I've spent too much money on useless plugins over the years so if a nice librarian/controller  is available for a reverbs or synth I invested in,  I'd def buy it. There's no such thing as boring protection , codes are easy to redeem. Even ilok licenses on the key were never a problem for me, But this is my take on it an not shared by everyone.
My suggestion is, if you can offer a software that is designed for popular gears for a relevant price, and you are willing to spend countless hours to maintain it, go for it. It requires C++ skills to update your juce application and a good understanding of ctrlr and LUA. Also you'll need some skills in desgn and marketing for the sale part.

Protecting a ctrlr based plugin is 100% feasible. As always, it deters hacker but is never fool proof. This at any level you want, and it's probably the easiest thing to do compared to the other efforts you'll have to do to release a good product.

As it is, ctrlr panels are not "commercial ready" at all, it's just xml file with a bunch of resources in a zip. If you want to secure it like your favorite plugns, you need to export as restricted instance with some customization, you'll need to encrypt the custom part of your code with the juce encryption module, you need to wrap the binary with obfuscating code (like with the pace wrapper), implement a licensing code redemption method, then to finish, prepare installers, then code sign the final product, as vst3, AU, application, for PC and Mac intel and mac silicon, you can also do it for ios and linux. It means 9-10 builds. That's a lot of extra work on top of just designing a well build midi panel... 

My advice would be to start with a very simple project like to control a small synth or fx with only CC involved (behringer etc), go all the way with the design, go through all the steps of protection and notarization and see if you can scale it to more demanding projects. Minimum time spent will be about 2 months full time, all included.

So good luck with your project, we are all here to help but get ready to spend a lot of time and effort, designing a panel will only be about 20-30% of the project if commercial, and count 20% about the protection+notarization of it .
Damien

EDIT : Or you can go the donationware route which seems a  very nice compromise to me.

**Reply by bijlevel** [reply]
I agree with the donationware route, which I'm considering myself for the panel I'm working on at the moment, because of the enormous amount of work e.q. hours I have, and still have to, put in.

**Comment by FabQi** [comment]
Hello

Thank you all for sharing your knolowede and point of view :-).
It's confirming what I knew and thought about it. 
(In an other life, I've been a pro IT guy for 25 years. Need a lot of update (and upgrade ?))

Anecdote :
French people, we LOVE resquiller (get in for free).
I love the ethic concept of donationware too .  But for me, it (almost) means free of charge. (not proud)

So, it's not a 'natural' path for me. But, why not...

Again, thank you for your replys 
Regards 
Fabrice

**Comment by Godlike-Productions** [comment]
I protect mine with PACE (iLok).  There is a fee to be paid, but it's affordable, and their protection is pretty good and full featured, and better than anything I could code.

**Comment by oldgearguy** [comment]
I'm at the point where I'd like to encrypt/protect a stand-alone version of my panel.
The Windows bundling seems fine.
I see that between Damien's posts and all the info from @monophreak I can tackle the Mac version.

Has there been any effort to create a bundled/encrypted version for Linux?

Also - to be compliant with the various licenses, what is needed and where should it go?  I will be distributing a user manual with the panel.

**Comment by damiensellier** [comment]
Hi,
I'm on vacations but when I'll get some time I'll write a whole tutorial about securing panels and exported instances. There are different levels of security you can apply to your designs from a simple passcode sent to the buyer to disable a top layer and unlocking LUA scripts for exemple, to the most advanced method such as implementing PACE iLok wrapping.

encrypting a panel is something else, it prevents people bringing the panel and its resources to a stock CtrlrX binary. The way the macOS binary is working is that if on load it detects some files in the bundle resources folder of the app, it loads it and work as a standalone export/plugin. 

Encrypting (JUCE blowfish) relies on the developper changing the encryption passkeys in CtrlrMac.cpp with a custom one then compile CtrlrX before loading the panel and exporting the instance, so that the resources and bundled panel cannot be copy pasted to the stock CtrlrX binaries. It's necessary but not sufficient.

https://github.com/damiensellier/CtrlrX/blob/fef3b8d86c7308974bba65a51b6e48be800e4a7b/Source/Native/CtrlrMac.cpp#L210
https://github.com/damiensellier/CtrlrX/blob/fef3b8d86c7308974bba65a51b6e48be800e4a7b/Source/Native/CtrlrMac.cpp#L897

I'll go deeper about all that during the summertime because it's a crucial topic.

**Comment by oldgearguy** [comment]
Thanks for the reply.   Speaking for myself,  I'd mainly be interested in making sure the source wasn't easily accessible.  Less need to tie a release to particular person/key.  The secondary request would be a way to issue a secure update (bug fixes,  features) to the folks that already have the original.

**Reply by damiensellier** [reply]
The source won't be accessible as soon as you compile CtrlrX with a custom blowfish encryption passkey on macOS. On windows it's a bit different, the panel and the resources are embedded in the exe binary which makes it harder to detach from the program.

Download CtrlrX, change the 2x bf passkey in CtrlrMac.cpp and compile it.  When you open a panel with this custom program and export a standalone app or plugin it, the panel and its resources are going to be encrypted with those custom blowfish passkeys which means that even a smart user cannot copy those resources from the bundle and insert them to a stock CtrlrX version. They would need YOUR custom CtrlrX version to decrypt it. CtrlrX in its stock version won't be able to open them and will start the app/plugin with a blank canvas.


I thought about encryting the resources as well for windows but I don't know if it's useful especially if after this step you run it through the pace ilok wrapping tool which obfuscate and scramble the actual code. it would probably be useless in this case.

To do online version check you can use the juce event module which is already pre installed and network classes are also bind to lua. You would need to run a script triggered when the panel is loaded to check the current panel version and check online if it's up to date with the latest declared version online.

It requires on your side to host a little code on a github repo or another host with a json showing the latest version number : 

```
{
  "latest_version": "1.2.4",
  "download_url": "https://yourwebsite.com/downloads"
}
```

then your script would look a bit like this, it's a sample from gemini so it does not hold the real version properties etc : 


```

-- =========================================================================
-- Callback triggered automatically by the C++ host when the panel finishes loading
-- =========================================================================
function onPanelLoaded(panelInstance)
    -- It's good practice to log or print to your host's console for debugging
    console("Panel loaded successfully. Initiating background version check...")

    -- 1. Define the remote URL holding your version manifest
    local versionUrl = "https://yourwebsite.com/version_check.json"
    
    -- 2. Grab the current plugin version from the host panel instance
    -- (Assuming panelInstance:getCurrentVersion() is bound via luabind)
    local currentVersion = panelInstance:getCurrentVersion() -- use getPropertyInt() with panelVersionMajor etc
    console("Current local version: " + currentVersion)

    -- 3. Trigger an asynchronous HTTP GET request.
    -- We pass a Lua callback function that will execute when the server responds.
    NetworkManager.fetchUrlAsync(versionUrl, function(success, responseText)
        
        -- Check if the HTTP request actually succeeded
        if not success then
            console("Version check failed: Unable to connect to server.")
            return
        end

        -- 4. Parse the JSON response
        -- (Assuming a JSON helper utilities library is bound to Lua)
        local data = JSON.parse(responseText)
        if not data or not data.latest_version then
            console("Version check failed: Invalid JSON response from server.")
            return
        end

        local latestVersion = data.latest_version
        local downloadUrl = data.download_url or "https://yourwebsite.com"

        -- 5. Version Comparison & Action
        if latestVersion ~= currentVersion then
            console("New version found: " .. latestVersion)
            
            -- Trigger a UI alert or notify the panel UI to display an "Update" button
            panelInstance:setUpdateAvailable(true, latestVersion, downloadUrl)
            
            -- Optional: Prompt a native host popup if your framework supports it
            HostUI.showAlertDialog(
                "Update Available",
                "Version " .. latestVersion .. " is now available! Would you like to download it?",
                downloadUrl
            )
        else
            console("Plugin is up to date.")
            panelInstance:setUpdateAvailable(false, currentVersion, "")
        end
    end)
end
```


don't look to much into the lua script it's just there to highlight the different steps with function placeholders but you'll get an idea.

Hope it helps

**Comment by oldgearguy** [comment]
Wow,  great information.   That's more than enough to get me started for now.   Enjoy the rest of your vacation.

**Reply by damiensellier** [reply]
you're welcome! Don't hesitate if ever you come across some issues. Maybe a required c++ luabind or a helper function is required for the online version check but I don't believe so, the JSON parsing class and InputStream should be enough.
